File: commands/scripts/prune_records.py
"""Scan / prune low-quality records in the estimator history CSV.

Reuses build_csv's parser/HEADERS/writer.
- detect <history.csv>: print {"prune":[{row,key,reasons}],"flag":[{row,key,warnings}]}
- apply <history.csv> <rows.json {"rows":[...]}>: delete those 1-based data rows,
  print {"deleted_rows":[...],"deleted_keys":[...],"row_count":N}
"""

import json
import os
import re
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "scripts"))
import build_csv  # noqa: E402

FIELD_KEYS = build_csv.FIELD_KEYS
COL = {fk: i for i, fk in enumerate(FIELD_KEYS)}
VALID_BASIS = {"actual", "final", "suggested"}


def _cell(row, fk):
    i = COL[fk]
    return row[i] if i < len(row) else ""


def _blank(value):
    return value is None or str(value).strip() == ""


_NUMBER_RE = re.compile(r"^\d+(\.\d+)?$")


def _is_number(value):
    return bool(_NUMBER_RE.match(str(value).strip()))


def detect(rows):
    prune = []
    flag = []
    for i, row in enumerate(rows, start=1):
        key = _cell(row, "key")
        reasons = []
        if _blank(key):
            reasons.append("missing key")
        if _blank(_cell(row, "summary")):
            reasons.append("missing summary")
        if not _is_number(_cell(row, "story_points")):
            reasons.append("missing/invalid story_points")
        if reasons:
            prune.append({"row": i, "key": key, "reasons": reasons})
            continue
        warnings = []
        if _cell(row, "estimate_basis").strip() not in VALID_BASIS:
            warnings.append("blank/invalid estimate_basis")
        if _blank(_cell(row, "issue_type")):
            warnings.append("missing issue_type")
        if warnings:
            flag.append({"row": i, "key": key, "warnings": warnings})
    return {"prune": prune, "flag": flag}


def apply_rows(rows, row_numbers):
    for n in row_numbers:
        if isinstance(n, bool) or not isinstance(n, int) or n < 1:
            raise ValueError("row numbers must be positive integers, got: %r" % (n,))
    delete_set = set(row_numbers)
    result = []
    deleted_rows = []
    deleted_keys = []
    for i, row in enumerate(rows, start=1):
        if i in delete_set:
            deleted_rows.append(i)
            deleted_keys.append(row[0] if row else "")
        else:
            result.append(row)
    return {"rows": result, "deleted_rows": deleted_rows, "deleted_keys": deleted_keys}


def _emit(obj):
    sys.stdout.buffer.write(
        (json.dumps(obj, ensure_ascii=False, separators=(",", ":")) + "\n").encode("utf-8")
    )


def main(argv):
    # argv is sys.argv: argv[1] is mode, argv[2] is history.csv, argv[3] is rows.json
    if len(argv) < 3:
        sys.exit("Usage: prune_records.py <detect|apply> <history.csv> [rows.json]")
    mode, history_csv = argv[1], argv[2]
    rows = build_csv.read_existing(history_csv)
    if mode == "detect":
        _emit(detect(rows))
    elif mode == "apply":
        if len(argv) != 4:
            sys.exit("Usage: prune_records.py apply <history.csv> <rows.json>")
        with open(argv[3], encoding="utf-8") as f:
            row_numbers = json.load(f).get("rows") or []
        try:
            result = apply_rows(rows, row_numbers)
        except ValueError as e:
            sys.exit("error: %s" % e)
        build_csv.write_csv(history_csv, result["rows"])
        _emit({
            "deleted_rows": result["deleted_rows"],
            "deleted_keys": result["deleted_keys"],
            "row_count": len(result["rows"]),
        })
    else:
        sys.exit("unknown mode: %s" % mode)


if __name__ == "__main__":
    main(sys.argv)

File: commands/scripts/prune_records.mjs
// Scan / prune low-quality records in the estimator history CSV.
// Reuses build_csv's parser/HEADERS/writer.
//   detect <history.csv> → {"prune":[{row,key,reasons}],"flag":[{row,key,warnings}]}
//   apply  <history.csv> <rows.json {"rows":[...]}> → {"deleted_rows":[...],"deleted_keys":[...],"row_count":N}
import { HEADERS, FIELD_KEYS as _FIELD_KEYS, readExisting, writeCsv } from "../../scripts/build_csv.mjs";
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

export const FIELD_KEYS = _FIELD_KEYS;
const COL = Object.fromEntries(FIELD_KEYS.map((fk, i) => [fk, i]));
const VALID_BASIS = new Set(["actual", "final", "suggested"]);

const cell = (row, fk) => (COL[fk] < row.length ? row[COL[fk]] : "");
const blank = (v) => v === null || v === undefined || String(v).trim() === "";
const NUMBER_RE = /^\d+(\.\d+)?$/;
const isNumber = (v) => NUMBER_RE.test(String(v).trim());

export function detect(rows) {
  const prune = [];
  const flag = [];
  rows.forEach((row, idx) => {
    const i = idx + 1;
    const key = cell(row, "key");
    const reasons = [];
    if (blank(key)) reasons.push("missing key");
    if (blank(cell(row, "summary"))) reasons.push("missing summary");
    if (!isNumber(cell(row, "story_points"))) reasons.push("missing/invalid story_points");
    if (reasons.length) { prune.push({ row: i, key, reasons }); return; }
    const warnings = [];
    if (!VALID_BASIS.has(cell(row, "estimate_basis").trim())) warnings.push("blank/invalid estimate_basis");
    if (blank(cell(row, "issue_type"))) warnings.push("missing issue_type");
    if (warnings.length) flag.push({ row: i, key, warnings });
  });
  return { prune, flag };
}

export function applyRows(rows, rowNumbers) {
  for (const n of rowNumbers) {
    if (!Number.isInteger(n) || n < 1) throw new Error(`row numbers must be positive integers, got: ${n}`);
  }
  const deleteSet = new Set(rowNumbers);
  const result = [];
  const deletedRows = [];
  const deletedKeys = [];
  rows.forEach((row, idx) => {
    const i = idx + 1;
    if (deleteSet.has(i)) { deletedRows.push(i); deletedKeys.push(row && row.length ? row[0] : ""); }
    else result.push(row);
  });
  return { rows: result, deleted_rows: deletedRows, deleted_keys: deletedKeys };
}

export function main(argv) {
  // argv is already sliced (process.argv.slice(2)): [mode, historyCsv, rowsJson]
  if (argv.length < 2) {
    console.error("Usage: prune_records.mjs <detect|apply> <history.csv> [rows.json]");
    process.exit(1);
  }
  const [mode, historyCsv] = argv;
  const rows = readExisting(historyCsv);
  if (mode === "detect") {
    console.log(JSON.stringify(detect(rows)));
  } else if (mode === "apply") {
    if (argv.length !== 3) { console.error("Usage: prune_records.mjs apply <history.csv> <rows.json>"); process.exit(1); }
    const rowNumbers = JSON.parse(readFileSync(argv[2], "utf8")).rows || [];
    let result;
    try { result = applyRows(rows, rowNumbers); }
    catch (e) { console.error(`error: ${e.message}`); process.exit(1); }
    writeCsv(historyCsv, result.rows);
    console.log(JSON.stringify({ deleted_rows: result.deleted_rows, deleted_keys: result.deleted_keys, row_count: result.rows.length }));
  } else {
    console.error(`unknown mode: ${mode}`);
    process.exit(1);
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main(process.argv.slice(2));
}
